### ISSUE TITLE
generic `Callables` as function arguments

### ISSUE THREAD

**Issue by iyakushev** [DetachHead/basedpyright#1005]
Hi! Looks like a bug to me.

Consider this example:

```python
def foo() -> str:
    return ""

def bar() -> int:
    return 42

def buz[**P, R](x: Callable[P, R], y: Callable[P, R]) -> R:
    ...

buz(foo, bar)  # No error here
```

Given two different functions `foo` and `bar` and a third function `buz`, which takes 2 fn as inputs, I expect that the type of `Callable[P, R]` will be inferred upon call and will result in a check for the same function signature between arguments `x` and `y`. But static analysis reports that everything is fine here. Obviously `foo != bar` since they diverge in a return type `R`.

**Comment by KotlinIsland** [comment]
`R` is inferred as `int | str`

**Comment by DetachHead** [comment]
yeah i don't think this is a bug, it's just how the generic is being inferred. ideally there would be a way to specify the generics explicitly when calling functions, but for some reason the python maintainers still haven't thought to add runtime support for it even though you can do it on classes
